src/include/graph/on_disk_graph.h

```diff
@@ -13,10 +22,54 @@ class MemoryManager;
 namespace graph {
 
 struct OnDiskGraphScanState {
-    std::unique_ptr<storage::RelTableScanState> fwdScanState;
-    std::unique_ptr<storage::RelTableScanState> bwdScanState;
-
-    explicit OnDiskGraphScanState(const storage::RelTable& table,
+    class InnerIterator {
+    public:
+        InnerIterator(const main::ClientContext* context, storage::RelTable* relTable,
+            std::unique_ptr<storage::RelTableScanState> tableScanState);
+
+        DELETE_COPY_DEFAULT_MOVE(InnerIterator);
+
+        std::span<const common::nodeID_t> getNodes() const {
+            // When scanning a single node, all returned results should be consecutive
+            // TODO(bmwinger): This won't be the case when we scan multiple
+            auto firstElement = dstSelVector().getSelectedPositions()[0];
+            RUNTIME_CHECK(for (size_t i = 0; i < dstSelVector().getSelSize(); i++) {
+                KU_ASSERT(firstElement + i == dstSelVector().getSelectedPositions()[i]);
+            });
+            return std::span<const common::nodeID_t>(
+                &dstVector().getValue<const common::nodeID_t>(firstElement),
+                dstSelVector().getSelSize());
+        }
+        std::span<const common::nodeID_t> getEdges() const {
+            // When scanning a single node, all returned results should be consecutive
+            auto firstElement = dstSelVector().getSelectedPositions()[0];
+            RUNTIME_CHECK(for (size_t i = 0; i < dstSelVector().getSelSize(); i++) {
+                KU_ASSERT(firstElement + i == dstSelVector().getSelectedPositions()[i]);
+            });
+            return std::span<const common::nodeID_t>(
+                &relIDVector().getValue<const common::nodeID_t>(firstElement),
+                dstSelVector().getSelSize());
+        }
+
+        bool next();
+        void initScan();
+
+    private:
+        const common::SelectionVector& dstSelVector() const {
```

Comment:
Can we also unify `dst` to `nbr`?

Reply:
I'd put dstSelVector there since it's the selection vector for the dstNodeIDVector. Or do you mean make the dstVectors be nbrVectors too? I guess that there is no longer a single output/destination storage with the edges being stored too, so that would probably be more appropriate.